From 858792f75fa8e2060360bed300bc55c2a2b8baa3 Mon Sep 17 00:00:00 2001
From: Jack Ping CHNG <jack.ping.chng@linux.intel.com>
Date: Sat, 4 Jul 2020 14:04:45 +0800
Subject: [PATCH] DRV_DPM_SW-337: DPM change synchronize_rcu to rcu_callback in
 dp_swdev

Should not use synchronize_rcu in spinlock context
---
 drivers/net/datapath/dpm/datapath_swdev.c | 11 ++++++++---
 drivers/net/datapath/dpm/datapath_swdev.h |  1 +
 2 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_swdev.c b/drivers/net/datapath/dpm/datapath_swdev.c
index 2aee056e65a9..be9384d34d57 100644
--- a/drivers/net/datapath/dpm/datapath_swdev.c
+++ b/drivers/net/datapath/dpm/datapath_swdev.c
@@ -133,15 +133,20 @@ struct br_info *dp_swdev_bridge_entry_lookup_rcu(struct net_device *dev)
 	return NULL;
 }
 
+static void dp_swdev_rcu_free(struct rcu_head *head)
+{
+	struct br_info *br_item = container_of(head, struct br_info, rcu_head);
+
+	kmem_cache_free(cache_br_info_list, br_item);
+}
+
 static void dp_swdev_remove_bridge_id_entry(struct br_info *br_item)
 {
 	DP_DEBUG(DP_DBG_FLAG_SWDEV, "Hash del(%s)\n", br_item->dev->name);
 
 	hlist_del_rcu(&br_item->br_hlist);
 
-	synchronize_rcu();
-
-	kmem_cache_free(cache_br_info_list, br_item);
+	call_rcu(&br_item->rcu_head, dp_swdev_rcu_free);
 }
 
 static void dp_swdev_insert_bridge_id_entry(struct br_info *br_item)
diff --git a/drivers/net/datapath/dpm/datapath_swdev.h b/drivers/net/datapath/dpm/datapath_swdev.h
index 1b0ca6f36869..a97c8ae1b544 100644
--- a/drivers/net/datapath/dpm/datapath_swdev.h
+++ b/drivers/net/datapath/dpm/datapath_swdev.h
@@ -46,6 +46,7 @@ struct br_info {
 	u32 num_vlan;			/* Num of vlans supported in br */
 	u32 num_fid;			/* Num of vlan aware fid in br */
 	struct list_head br_vlan_list;  /* List pointing to the vlan entry */
+	struct rcu_head rcu_head;
 };
 
 struct bridge_member_port {
